### PR TITLE
[CALCITE-2223][CALCITE-2589] Avoid infinite match of the rules when relation inputs form a cycle

### PR DESCRIPTION
**Pull request by vlsi**
https://issues.apache.org/jira/browse/CALCITE-2223
https://issues.apache.org/jira/browse/CALCITE-2589

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/SortRemoveConstantKeysRule.java [hunk 1]
```diff
@@ -71,15 +71,15 @@ private SortRemoveConstantKeysRule() {
 
     // No active collations. Remove the sort completely
     if (collationsList.isEmpty() && sort.offset == null && sort.fetch == null) {
-      call.transformTo(input);
       call.getPlanner().setImportance(sort, 0.0);
+      call.transformTo(input);
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rel/rules/SortRemoveConstantKeysRule.java]: Why does the order matter?

File: core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java [hunk 2]
```diff
@@ -537,6 +540,11 @@ protected RelOptPlanner createPlanner(
       planner.addRule(AbstractMaterializedViewRule.INSTANCE_JOIN);
       planner.addRule(AbstractMaterializedViewRule.INSTANCE_PROJECT_AGGREGATE);
       planner.addRule(AbstractMaterializedViewRule.INSTANCE_AGGREGATE);
+    } else {
+      // See [CALCITE-2223]. Enabling the rules with materializations might result in infinite
+      // planning. E.g. org.apache.calcite.test.MaterializationTest.testJoinMaterialization10
+      planner.addRule(FilterSetOpTransposeRule.INSTANCE);
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java]: @julianhyde , I'm not sure if those rules should be activated by default, however I think we can do that.
Of course `FilterSetOpTransposeRule` does not currently play well with materializations, so I include certain rules under `if (not materializations_enabled)`

Objections?

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java [hunk 3]
```diff
@@ -1350,7 +1350,6 @@ private void basePushFilterPastAggWithGroupingSets(boolean unchanged)
             + "(select * from emp e1 union all select * from emp e2) r2");
   }
 
-  @Ignore("cycles")
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: I've no idea what is the expected output for the tests, however those tests do not cycle anymore, so I activate them

**Comment by julianhyde** [comment]
There are quite a few "sensible" rules that cause problems for materialized view matching, e.g. pushing down projects. I think pushing a Filter into a Union is another such "sensible" rule, and we should enable by default.

When matching materialized views, people do usually-considered-crazy stuff like pulling up projects.

I do agree that our "default rules" are pretty unwieldy, and we should give people a better way to manage them. Perhaps organize them into "rule sets" or "programs".

**Comment by vlsi** [comment]
Frankly speaking, I've spend a day or so debugging `testJoinMaterialization10`, and I think it might happen too often in practice

File: spark/src/test/java/org/apache/calcite/test/SparkAdapterTest.java [hunk 4]
```diff
@@ -622,11 +622,13 @@
...
 
     final String plan = "PLAN="
-        + "EnumerableUnion(all=[false])\n"
-        + "  EnumerableCalc(expr#0..1=[{inputs}], expr#2=[1], expr#3=[>($t0, $t2)], X=[$t0], $condition=[$t3])\n"
-        + "    EnumerableValues(tuples=[[{ 1, 'a' }, { 2, 'b' }]])\n"
-        + "  EnumerableCalc(expr#0..1=[{inputs}], expr#2=[1], expr#3=[>($t0, $t2)], X=[$t0], $condition=[$t3])\n"
-        + "    EnumerableValues(tuples=[[{ 1, 'a' }, { 2, 'b' }, { 1, 'b' }, { 2, 'c' }, { 2, 'c' }]])\n\n";
+        + "EnumerableAggregate(group=[{0}])\n"
+        + "  EnumerableUnion(all=[true])\n"
+        + "    EnumerableCalc(expr#0..1=[{inputs}], expr#2=[1], expr#3=[>($t0, $t2)], X=[$t0], $condition=[$t3])\n"
```
Comment (vlsi) [inline spark/src/test/java/org/apache/calcite/test/SparkAdapterTest.java]: This looks a bit odd since it does not pre-group the first input.

**Pull request closed without merging**